src/org/sosy_lab/cpachecker/cfa/parser/eclipse/c/ASTConverter.java

```diff
@@ -2579,47 +2810,117 @@ private CEnumType convert(IASTEnumerationSpecifier d) {
   }
 
+  private static final ImmutableList<CSimpleType> PACKED_ENUM_REPRESENTATION_CANDIDATE_TYPES =
+      ImmutableList.of( // list of types with incrementing size, less than int
+          CNumericTypes.CHAR,
+          CNumericTypes.UNSIGNED_CHAR,
+          CNumericTypes.SHORT_INT,
+          CNumericTypes.UNSIGNED_SHORT_INT);
+
   private static final ImmutableList<CSimpleType> ENUM_REPRESENTATION_CANDIDATE_TYPES =
       ImmutableList.of( // list of types with incrementing size
-          CNumericTypes.SIGNED_INT, CNumericTypes.UNSIGNED_INT, CNumericTypes.SIGNED_LONG_LONG_INT);
+          CNumericTypes.SIGNED_INT,
+          CNumericTypes.UNSIGNED_INT,
+          CNumericTypes.SIGNED_LONG_INT, // XXX why were no longs
+          CNumericTypes.UNSIGNED_LONG_INT,
+          CNumericTypes.SIGNED_LONG_LONG_INT,
+          CNumericTypes.UNSIGNED_LONG_LONG_INT);
 
   /**
-   * Compute a matching integer type for an enumeration. We use SIGNED_INT and switch to larger type
-   * if needed.
+   * Compute a matching integer type for each enumerator and for enumeration.
    *
    * <p>§6.7.2.2 (4) Each enumerated type shall be compatible with char, a signed integer type, or
    * an unsigned integer type. The choice of type is implementation-defined, but shall be capable of
    * representing the values of all the members of the enumeration.
+   *
+   * <p>GCC allows enumerators with values out of <code>int</code> bounds (breaks §6.7.2.2 (2)). GCC
+   * allows enums to be packed (using <code>__packed__</code> attribute).
+   *
+   * <p>For all enumerators and not packed enums GCC selects smallest type from (signed/unsigned)
+   * int, long, long long. For packed enums GCC selects from all types staring with (signed/unsigned)
+   * char and ending with long long.
+   *
+   * <p>Note that types of enumerators are not the same as type of the enum and also may differ from
+   * each other.
+   *
+   * @return integer type compatible with this enum
    */
-  private CSimpleType getEnumerationType(final CEnumType enumType) {
-    LongSummaryStatistics enumStatistics =
-        enumType.getEnumerators().stream()
-            .filter(CEnumerator::hasValue) // some values might not have been simplified
-            .mapToLong(CEnumerator::getValue)
-            .summaryStatistics();
-
-    Preconditions.checkState(
-        enumStatistics.getCount() > 0, "enumeration does not provide any values: %s", enumType);
-    final BigInteger minValue = BigInteger.valueOf(enumStatistics.getMin());
-    final BigInteger maxValue = BigInteger.valueOf(enumStatistics.getMax());
+  private CSimpleType getEnumerationType(boolean pPacked, List<CEnumerator> pEnumerators) {
+    BigInteger minValue = null;
+    BigInteger maxValue = null;
+
+    for (CEnumerator enumerator : pEnumerators) {
+      if (!enumerator.hasValue()) {
+        // some values might not have been simplified
+        // well, let it be int then
+        enumerator.setType(CNumericTypes.SIGNED_INT);
```

Comment:
Please investigate whether this case is possible. Also please check if c.ASTConverter.convert(IASTEnumerationSpecifier.IASTEnumerator e, Long lastValue) correctly works especially with values close to ULLONG_MAX

Reply:
About ULLong values: It was not working correctly, but now I replaced long value field with BigInteger where needed.
I will check in which case enumerator does not have value later.

Reply:
About not simplified enumerator values: this case was possible; i hope i fixed it.